### PR TITLE
Add audible messages to confirm the "move block" actions

### LINKED ISSUE #23995: Add audible messages to confirm the "move block" actions.
**Describe the bug**

#23276 added keyboard shortcuts to move the blocks. Nice improvement for all keyboard users. Pending #3218.

However, screen reader users don't get any confirmation of the move action. The only chance they have to check a block has been moved to the intended position is to explore the content around the block and verify it in a tedious, laborious, process.

It would be great to add a couple simple audible `speak()` messages to confirm the result of the move actions.

ideally the messages should also announce the new position. For example:
- "Block moved up to position {n}"
- "Block moved down to position {n}"

### PR DESCRIPTION
**Pull request by enricobattocchi**
Fixes #23995.

## Description

Added audible messages to `aria-live="assertive"` region to confirm the "move" action to the screen reader users.
The messages check for orientation and RTL to make sure they are consistent with the label of the button used to perform the action.

_This PR has been created during the Yoast Contributor Day.
The people who worked on it are  @jcomack, @afercia, @enricobattocchi._

## How has this been tested?

1. edit a (new) post with a screen reader active
2. add a number of blocks, plus a button list (which has a `horizontal` orientation)
3. move the blocks using the buttons or the keyboard shortcuts and check that the actions are confirmed by the audible messages. Notice that button can be moved left or right and the messages change accordingly.
4. Try also to select multiple blocks and move them together to hear a message telling the correct number of blocks moved.
5. Switch to a RTL environment and see that the "move up" action becomes "move right" for the buttons, and the audible messages are correct.
6. You can also the content of the `<div id="a11y-speak-assertive">` to read the messages without a screen reader active.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by afercia** [comment]
To fully test, worth reminding the keyboard shortcuts to move blocks are (on macOS):
- Shift Control Option Y: move down (or right for LTR)
- Shift Control Option T: move up (or left for LTR) 

Moving left or right can be tested on the "Buttons" or "Social icons" blocks:

Screenshot 2020-08-29 at 16 27 33

Screenshot illustrating moving three social icons buttons when using Safari and VoiceOver:

Screenshot 2020-08-29 at 16 27 07

File: packages/block-editor/src/store/effects.js [hunk 1]
```diff
@@ -219,6 +271,30 @@ export default {
...
+
+		let direction = 'down';
+
+		if ( action.orientation === 'horizontal' ) {
+			direction = settings.isRTL ? 'left' : 'right';
+		}
+
+		announceBlockMoved( state, direction );
+	},
+	MOVE_BLOCKS_UP: ( action, { getState } ) => {
```
Comment (mtias) [inline packages/block-editor/src/store/effects.js:286]: It's odd to read an action type of "UP" while orientation changes it to right / left within. It'd be better to have a single "MOVE_BLOCKS" type and handle all cardinals within.

File: packages/block-editor/src/store/effects.js [hunk 2]
```diff
@@ -219,6 +271,30 @@ export default {
...
 		);
 	},
+	MOVE_BLOCKS_DOWN: ( action, { getState } ) => {
+		const state = getState();
+		const settings = getSettings( state );
+
+		let direction = 'down';
+
+		if ( action.orientation === 'horizontal' ) {
+			direction = settings.isRTL ? 'left' : 'right';
```
Comment (mtias) [inline packages/block-editor/src/store/effects.js:281]: Is this correct? Wouldn't the actual move actions still be accurate given left moves to the left regardless of RTL?

File: packages/block-editor/src/store/effects.js [hunk 3]
```diff
@@ -66,6 +67,57 @@ export function validateBlocksToTemplate( action, store ) {
...
+ * Announces when a block is moved.
+ *
+ * @param {Object} state The current state.
+ * @param {string} direction The direction the block was moved in.
+ */
+export function announceBlockMoved( state, direction ) {
+	const blockCount = getSelectedBlockCount( state );
+	let message;
+
+	switch ( direction ) {
```
Comment (mtias) [inline packages/block-editor/src/store/effects.js:80]: It seems all these could be inlined and don't need the extra abstraction of `announceBlockMoved`

File: packages/block-editor/src/store/effects.js (hunk 1 above)
Comment (afercia) [reply packages/block-editor/src/store/effects.js:286]: This just uses the existing actions. Actually, there are no specific actions for "left" and "right". Not opposed to improvements but the current actions implementation is out of the scope of this issue.

File: packages/block-editor/src/store/effects.js (hunk 2 above)
Comment (afercia) [reply packages/block-editor/src/store/effects.js:281]: It is correct. "down" moves the block to the "following" position, whether it's vertical or horizontal. In LTR, the "following" horizontal position is right, while in RTL is left.

File: packages/block-editor/src/store/effects.js (hunk 3 above)
Comment (afercia) [reply packages/block-editor/src/store/effects.js:80]: I'd disagree. Seems to me it's way more readable this way.

**Comment by afercia** [comment]
I tested again this PR, also on Windows Firefox + NVDA and in RTL. Everything works as expected. Code looks good as is, to me. Kindly asking to have a new review please, when there's a chance 🙂 /Cc @youknowriad @talldan 

In RTL: last action was "move to right":

Screenshot (127)

File: packages/block-editor/src/store/effects.js [hunk 4]
```diff
@@ -219,6 +271,30 @@ export default {
 			)
 		);
 	},
+	MOVE_BLOCKS_DOWN: ( action, { getState } ) => {
```
Comment (youknowriad) [inline packages/block-editor/src/store/effects.js:274]: Can we move away from effects and instead implement these action creator as "generators"?  see `moveBlocksToPosition` as an example of action generator.

Comment (afercia) [reply packages/block-editor/src/store/effects.js:274]: Not sure I follow. Maybe I lack context or it's a terminology thing 🙂 In this PR we just reused actions that already exist. Are you suggesting a major refactoring in the way these actions are created?

Comment (youknowriad) [reply packages/block-editor/src/store/effects.js:274]: We're slowly moving away from using "effects" (see effects.js file and the https://github.com/aduth/refx middleware) and instead we're trying to replace them with "generator" actions (consistency with resolvers).

These are just two technologies that allow the same thing: writing complex actions with side effects. 
So ideally, new code shouldn't add new effects and use generators instead.

Comment (afercia) [reply packages/block-editor/src/store/effects.js:274]: Good to know. Is this documented anywhere? How new contributors are supposed to be informed about this? 🙂 If it's not documented, I'd say it's a barrier to new contributors...

Comment (youknowriad) [reply packages/block-editor/src/store/effects.js:274]: You're right, it should be. We were hoping we could get rid of these files which remove the confusion but the refactor efforts staled.

**Comment by afercia** [comment]
i created a new issue at https://github.com/WordPress/gutenberg/issues/67237 to recap the current situation.
@enricobattocchi although this PR can be used as source of inspiration, the codebase has changed significantly in the last 4 years. Do you think it can still be used and rebased or maybe better to start from scratch?

**Comment by enricobattocchi** [comment]
Thanks @afercia, I think it's better to start from scratch. I'll close this one.

**Pull request closed without merging**